virtual_ecosystem/models/abiotic/abiotic_model.py

```diff
@@ -0,0 +1,325 @@
+            atmospheric_pressure=self.data["atmospheric_pressure"].to_numpy()[0],
+            sensible_heat_flux_topofcanopy=(
+                self.data["sensible_heat_flux_topofcanopy"].to_numpy()
+            ),
+            wind_speed_ref=(
+                self.data["wind_speed_ref"].isel(time_index=time_index).to_numpy()
+            ),
+            wind_reference_height=(self.data["canopy_height"] + 10).to_numpy(),
+            abiotic_constants=self.model_constants,
+            core_constants=self.core_constants,
+        )  # TODO wind height above in constants, cross-check with reference heights
+
+        wind_output = {}
+        wind_speed_above_canopy = DataArray(
+            wind_update["wind_speed_above_canopy"],
+            dims="cell_id",
+            coords={"cell_id": self.data.grid.cell_id},
+        )
+        wind_output["wind_speed_above_canopy"] = wind_speed_above_canopy
+
+        for var in ["wind_speed_canopy", "friction_velocity"]:
+            var_out = DataArray(
+                np.concatenate(
+                    (
+                        wind_update[var][0 : (true_canopy_layers_n + 1)],
+                        np.full((empty_canopy_layers, self.data.grid.n_cells), np.nan),
+                        wind_update[var][
+                            (true_canopy_layers_n + 1) : (len(wind_update[var]) + 1)
+                        ],
+                        np.full(
+                            (
+                                len(self.layer_structure.soil_layers),
+                                self.data.grid.n_cells,
+                            ),
+                            np.nan,
+                        ),
+                    ),
+                ),
+                dims=self.data["layer_heights"].dims,
+                coords=self.data["layer_heights"].coords,
+            )
+            wind_output[var] = var_out
+
+        self.data.add_from_dict(output_dict=wind_output)
+
+        # Soil energy balance
+        topsoil_layer_index = self.layer_structure.layer_roles.index("soil")
+
+        soil_heat_balance = soil_energy_balance.calculate_soil_heat_balance(
+            data=self.data,
+            topsoil_layer_index=topsoil_layer_index,
+            update_interval=43200,  # TODO self.model_timing.update_interval
+            abiotic_consts=self.model_constants,
+            core_consts=self.core_constants,
+        )
+
+        soil_output = {}  # TODO add these variables to combined flux variables
+        var_list = [
+            "soil_absorption",
+            "longwave_emission_soil",
+            "sensible_heat_flux_soil",
+            "latent_heat_flux_soil",
+            "ground_heat_flux",
+        ]
+        for var in var_list:
+            soil_output[var] = DataArray(
+                soil_heat_balance[var],
+                dims="cell_id",
+                coords={"cell_id": self.data.grid.cell_id},
+            )
```

Comment:
Why is that faster @dalonsoa? I'd always understood dict comprehensions as being largely a clean syntax rather than a performance thing. This is still extending a dict, one step at a time?

I think it is cleaner - I just want to grok what the performance advantage is!

Reply:
That's a good point. List comprehensions are indeed faster (there are several posts about it) and I had assumed that it was the same for dictionaries. But it turns out it depends on the size of the dictionary, according to [this post](https://stackoverflow.com/questions/30050481/dictionary-comprehensions-in-python-are-too-comprehensive) (which is about python 2.7, so it can be massively outdated). All in all, there's not such a big difference, it would seem, so probably it comes down to readability choosing one or the other.